index.js

```diff
@@ -24,62 +43,88 @@ module.exports = {
+          const attachment = await self.db.findOne({ _id: _id });
+          if (!attachment) {
+            throw 'notfound';
+          }
+          if (attachment.utilized) {
 
-          // Once an attachment is attached to its first doc, the permissions
-          // of its docs become its permissions. The test is: we have to be able to view
-          // at least one of those docs as this user.
-          //
-          // Docs in the trash can pass this test for someone allowed to edit them,
-          // but in practice only come through if it happens to be the "allowed in
-          // trash" size used for media library preview; the rest
-          // will have permissions set to 000 or the disabledFileKey renaming pattern
-          // in effect, so things behave just as they would without this module.
+            // Once an attachment is attached to its first doc, the permissions
+            // of its docs become its permissions. The test is: we have to be able to view
+            // at least one of those docs as this user.
+            //
+            // Docs in the trash can pass this test for someone allowed to edit them,
+            // but in practice only come through if it happens to be the "allowed in
+            // trash" size used for media library preview; the rest
+            // will have permissions set to 000 or the disabledFileKey renaming pattern
+            // in effect, so things behave just as they would without this module.
 
-          const ids = (attachment.docIds || []).concat(attachment.trashDocIds || []);
-          if (!ids.length) {
-            throw 'forbidden';
+            const ids = (attachment.docIds || []).concat(attachment.trashDocIds || []);
+            if (!ids.length) {
+              throw 'forbidden';
+            }
+            const doc = await self.apos.docs.find(req, { _id: { $in: ids } }).trash(null).published(null).toObject();
+            if (!doc) {
+              // We are not cool enough to view any of the docs that
+              // contain this attachment (there can be more than one
+              // due to page copying, piece copying and workflow)
+              throw 'forbidden';
+            }
           }
-          const doc = await self.apos.docs.find(req, { _id: { $in: ids } }).trash(null).published(null).toObject();
-          if (!doc) {
-            // We are not cool enough to view any of the docs that
-            // contain this attachment (there can be more than one
-            // due to page copying, piece copying and workflow)
-            throw 'forbidden';
+        }
+        // OK, stream it, take care not to allow escape from the folder
+        path = path.replace(/\.\.\//g, '');
+        const contentType = contentTypes[extension] || 'application/octet-stream';
+        const input = self.uploadfs.streamOut(`/${path}`);
+        let progress = false;
+        try {
+          let n = 0;
+          for await (const buffer of input) {
```

Comment:
I looked at piping it, but this made it hard to know when I could and couldn't send an error message. This way I know when I've already started sending data and when I haven't